services/galley/src/Galley/API/Update.hs

```diff
@@ -439,9 +441,14 @@ withValidOtrBroadcastRecipients usr clt rcps val now go = Teams.withBindingTeam
     tMembers <- fmap (view userId) <$> Data.teamMembers tid
     contacts <- getContactList usr
     let users = Set.toList $ Set.union (Set.fromList tMembers) (Set.fromList contacts)
-    clts  <- Data.lookupClients users
+    clts  <- lookupClientsAsync users
     let membs = Data.newMember <$> users
     handleOtrResponse usr clt rcps membs clts val now go
+  where
+    lookupClientsAsync users = do
+        -- do a maximum of 16 parallel lookups of up to 128 users each
+        cltsList  <- forM (chunksOf 2048 users) $ \us -> mapConcurrently Data.lookupClients (chunksOf 128 us)
```

Comment:
why are we splitting into chunks of 2048 and then chunks of 128? it seems to me like we could just split in chunks of 128 directly?

Reply:
I think this is to limit each bunch of concurrent lookups to 16 running in parallel, instead of an arbitrary number running in parallel. Each chunk of 16 parallel lookups is done sequentially.

Reply:
> I think this is to limit each bunch of concurrent lookups to 16 running in parallel, instead of an arbitrary number running in parallel.

Right, that was the idea. Currently, this is limited by our model (1000 connections + 127 team members) but this tries to guarantee that this lookup would still have some "expectable" performance regardless of that limit.

Reply:
makes sense :+1: